src/core/binary.c

```diff
@@ -0,0 +1,201 @@
+/*
+* Copyright (c) 2023 Calvin Rose
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to
+* deal in the Software without restriction, including without limitation the
+* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+* sell copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+* IN THE SOFTWARE.
+*/
+
+#ifndef JANET_AMALG
+#include "features.h"
+#include <janet.h>
+#include "util.h"
+#endif
+
+static int binary_should_reverse(const Janet *argv, int32_t argc) {
+    JanetKeyword order_kw = janet_getkeyword(argv, argc);
+    if (!janet_cstrcmp(order_kw, "le")) {
+#if JANET_BIG_ENDIAN
+        return 1;
+#endif
+    } else if (!janet_cstrcmp(order_kw, "be")) {
+#if JANET_LITTLE_ENDIAN
+        return 1;
+#endif
+    } else if (!janet_cstrcmp(order_kw, "native")) {
+        return 0;
+    } else {
+        janet_panicf("expected endianness :le, :be or :native, got %v", argv[1]);
+    }
+    return 0;
+}
+
+static void binary_reverse_u32(uint8_t bytes[4]) {
+    uint8_t temp;
+    temp = bytes[3];
+    bytes[3] = bytes[0];
+    bytes[0] = temp;
+    temp = bytes[2];
+    bytes[2] = bytes[1];
+    bytes[1] = temp;
+}
+
+static void binary_reverse_u64(uint8_t bytes[8]) {
+    uint8_t temp;
+    temp = bytes[7];
+    bytes[7] = bytes[0];
+    bytes[0] = temp;
+    temp = bytes[6];
+    bytes[6] = bytes[1];
+    bytes[1] = temp;
+    temp = bytes[5];
+    bytes[5] = bytes[2];
+    bytes[2] = temp;
+    temp = bytes[4];
+    bytes[4] = bytes[3];
+    bytes[3] = temp;
+}
+
+JANET_CORE_FN(cfun_binary_write_uint16,
+              "(binary/write-uint16 buffer order data)",
+              "Write the binary representation of data into the buffer.") {
+    janet_fixarity(argc, 3);
+    JanetBuffer *buffer = janet_getbuffer(argv, 0);
+
+    int reverse = binary_should_reverse(argv, 1);
+
+    union {
+        uint16_t data;
+        uint8_t bytes[2];
+    } u;
+
+    u.data = (uint16_t) janet_getinteger(argv, 2);
```

Comment:
I was thinking this should be `janet_getuinteger` which exists below `janet_getinteger` in `src/core/capi.c` but I was getting these warnings/errors on `make valtest`:

```
cc -O2 -g -std=c99 -Wall -Wextra -Isrc/include -Isrc/conf -fvisibility=hidden -fPIC -c build/c/janet.c -o build/janet.o
src/core/binary.c: In function ‘cfun_binary_write_uint16’:
src/core/binary.c:86:25: warning: implicit declaration of function ‘janet_getuinteger’; did you mean ‘janet_getinteger’? [-Wimplicit-function-declaration]
   86 |     u.data = (uint16_t) janet_getuinteger(argv, 2);
      |                         ^~~~~~~~~~~~~~~~~
      |                         janet_getinteger
src/core/capi.c: At top level:
src/core/capi.c:296:10: error: conflicting types for ‘janet_getuinteger’; have ‘uint32_t(const Janet *, int32_t)’ {aka ‘unsigned int(const Janet *, int)’}
  296 | uint32_t janet_getuinteger(const Janet *argv, int32_t n) {
      |          ^~~~~~~~~~~~~~~~~
src/core/binary.c:86:25: note: previous implicit declaration of ‘janet_getuinteger’ with type ‘int()’
   86 |     u.data = (uint16_t) janet_getuinteger(argv, 2);
      |                         ^~~~~~~~~~~~~~~~~
make: *** [Makefile:223: build/janet.o] Error 1
```